File: ruby solutions/path_sum_ii.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @param {Integer} target_sum
# @return {Integer[][]}
def path_sum(root, target_sum)
    ans = []
    dfs(root, [], ans, target_sum)
    return ans
end

def dfs(root, path, ans, remaining_sum)
    return if root == nil
    
    path << root.val

    ans << path.dup if root.left == nil && root.right == nil && remaining_sum == root.val  

    dfs(root.left, path, ans, remaining_sum - root.val)
    dfs(root.right, path, ans, remaining_sum - root.val)

    path.pop
end

File: python solutions/path_sum_ii.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        ans = []
        self.dfs(root, [], ans, targetSum)
        return ans

    def dfs(self, root, path, ans, remainingSum):
        if not root:
            return

        path.append(root.val)
        if not root.left and not root.right and remainingSum == root.val:
            ans.append(list(path))

        self.dfs(root.left, path, ans, remainingSum - root.val)
        self.dfs(root.right, path, ans, remainingSum - root.val)

        path.pop()